docs(native): Add documentation for Presto C++ Installation (#26718)
prestodb/presto: presto-docs/src/main/sphinx/presto_cpp/installation.rst

=======================
Presto C++ Installation
=======================

> .. contents::

Reviewer: Look at the table of contents in a local doc build. (If you are not generating local doc builds, they are very helpful to see how your changes look before you push to GitHub.) 

See how Prerequisites is the only item in this table of contents at the top of the page, and then look at the table of contents in the right sidebar. You should probably raise up every heading in this page by one level. 
<img width="581" height="193" alt="Screenshot 2026-01-20 at 2 51 08 PM copy" src="https://github.com/user-attachments/assets/28a8b7b7-30ad-45fe-9b5d-9c50b031df69" />
<img width="250" height="630" alt="Screenshot 2026-01-20 at 2 51 08 PM" src="https://github.com/user-attachments/assets/7ce784d5-1d13-41f7-bea1-e36dde0096e8" />

1. Create Configuration Directory
"""""""""""""""""""""""""""""""""

To create the necessary directories for the coordinator and its catalogs, run the following command:

.. code-block:: bash

   mkdir -p coordinator/etc/catalog

2. Create the Coordinator Configuration File
""""""""""""""""""""""""""""""""""""""""""""

Create the file ``coordinator/etc/config.properties`` with the following contents. This file enables the coordinator mode, the discovery server, and sets the HTTP port to ``8080``.

.. code-block:: properties

   # coordinator/etc/config.properties
   coordinator=true
   node-scheduler.include-coordinator=true
   http-server.http.port=8080
   discovery-server.enabled=true
   discovery.uri=http://localhost:8080

* ``coordinator=true``: Enables the coordinator mode.
* ``discovery-server.enabled=true``: Designates the coordinator as the host for the worker discovery service.
* ``http-server.http.port=8080S``: Start the HTTP server on port 8080 for the coordinator (and workers, if enabled).

3. Create the JVM Configuration File
""""""""""""""""""""""""""""""""""""

> Create the file ``coordinator/etc/jvm.config`` with the following content. These are standard Java 17 flags for Presto. (It ensures compatibility with Java 17's module system, provide stable garbage collection, memory behavior, and enforce safe failure handling).

Reviewer: ```suggestion
Create the file ``coordinator/etc/jvm.config`` with the following content. These are standard Java 17 flags for Presto that ensures compatibility with Java 17's module system, provides stable garbage collection and memory behavior, and enforces safe failure handling.
```

4. Create the Node Properties File
""""""""""""""""""""""""""""""""""

Create the file ``coordinator/etc/node.properties`` with the following content. This file sets the node environment and the data directory.

.. code-block:: properties

   # coordinator/etc/node.properties
   node.id=${ENV:HOSTNAME}
   node.environment=test
   node.data-dir=/var/lib/presto/data

5. Create the TPC-H Catalog Configuration File
""""""""""""""""""""""""""""""""""""""""""""""

Create the file ``coordinator/etc/catalog/tpch.properties`` with the following content. The TPC-H catalog enables running test queries against an in-memory dataset.

.. code-block:: properties

   # coordinator/etc/catalog/tpch.properties
   connector.name=tpch

Configure the Prestissimo (C++) Worker
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Configure the Worker to locate the Coordinator or Discovery service and identify itself within the network.

1. Create Worker Configuration Directory
""""""""""""""""""""""""""""""""""""""""

.. code-block:: bash

   mkdir -p worker-1/etc/catalog

2. Create ``worker-1/etc/config.properties``
""""""""""""""""""""""""""""""""""""""""""""

> Configure the worker to point to the discovery service running on the coordinator. Repeat this step to add more workers, such as ``worker-2``.

Reviewer: ```suggestion
Configure the worker to point to the discovery service running on the coordinator. 

Note: You can repeat this step to add more workers, such as ``worker-2``.
```
I changed the phrasing because I think adding more workers is optional, not required.

.. code-block:: properties

   # worker-1/etc/config.properties
   discovery.uri=http://coordinator:8080
   presto.version=0.288-15f14bb
   http-server.http.port=7777
   shutdown-onset-sec=1
   runtime-metrics-collection-enabled=true

* ``discovery.uri=http://coordinator:8080``: This uses the coordinator service name as defined in the ``docker-compose.yml`` file for network communication within Docker.

3. Configure ``worker-1/etc/node.properties``
"""""""""""""""""""""""""""""""""""""""""""""

Define the worker’s internal address to ensure reliable registration.

.. code-block:: properties

   # worker-1/etc/node.properties
   node.environment=test
   node.internal-address=worker-1
   node.location=docker
   node.id=worker-1

> * ``node.internal-address=worker-1``: This setting matches the service name defined in Docker Compose.

Reviewer: Suggest replacing "Docker Compose" with a link to the heading "Create docker-compose.yml" in line 178.

=======================
Presto C++ Installation
=======================

.. contents::
    :local:
    :backlinks: none
    :depth: 1

> This shows how to install and run a lightweight Presto cluster utilizing a PrestoDB Java Coordinator and Prestissimo (C++) Workers using Docker.

Reviewer: ```suggestion
This shows how to install and run a lightweight Presto cluster utilizing a PrestoDB Java Coordinator and Prestissimo (Presto C++) Workers using Docker. 
For more information about Presto C++, see the :ref:`presto-cpp:overview`.
```

This new page in the Presto documentation does not have to explain what is Prestissimo, we can refer to the existing doc for that, and focus this page on how to deploy Presto with Prestissimo workers.

The setup uses Meta's high-performance Velox engine for worker-side query execution. We will configure a cluster and run a test query with the built-in TPCH connector.

> Introducing Prestissimo (Presto C++ Worker)

Reviewer: Please delete lines 14-23, as we are leveraging the existing doc with the new sentence in line 10.

Prestissimo is the C++ native implementation of the Presto :ref:`overview/concepts:worker`. It is designed to be a drop-in replacement for the traditional Java worker. It is built using Velox, a high-performance, open-source C++ database acceleration library created by Meta.

A C++ execution engine offers significant advantages for data lake analytics:

* **Massive Performance Boost:** Prestissimo achieves increases in CPU efficiency and reduces query latency by leveraging native C++ execution, vectorization, and SIMD (Single Instruction, Multiple Data) instructions.
* **Eliminates Java Garbage Collection Issues:** By moving the execution engine out of the Java Virtual Machine (JVM), this architecture removes performance spikes and pauses associated with Java Garbage Collection, resulting in more consistent and stable query times.
* **Explicit Memory Control:** The Velox memory management framework offers explicit memory accounting and arbitration, providing finer control over resource consumption than in the JVM.

Prerequisites
-------------

To follow this tutorial, you need:

* Docker installed.
* Basic familiarity with the terminal and shell commands.

Setup Guide
-----------

> The recommended directory structure uses ``presto-lab`` as the root directory.

Reviewer: Move this sentence to become the first sentence after the heading `Create a Working Directory`.

Create a Working Directory
^^^^^^^^^^^^^^^^^^^^^^^^^^

Create a clean root directory to hold all necessary configuration files and the ``docker-compose.yml`` file.

.. code-block:: bash

   mkdir -p ~/presto-lab
   cd ~/presto-lab

Configure the Presto Java Coordinator
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The Coordinator requires configuration to define its role, enable the discovery service, and set up a catalog for querying.

> A. Create Configuration Directory

Reviewer: ```suggestion
1. Create Configuration Directory
```

"""""""""""""""""""""""""""""""""

.. code-block:: bash

   mkdir -p coordinator/etc/catalog

> This command creates the necessary directories for the coordinator and its catalogs.

Reviewer: ```suggestion
To create the necessary directories for the coordinator and its catalogs, run the following command: 
```
Make this an imperative, and move it to before `mkdir -p coordinator/etc/catalog`.

B. Create ``coordinator/etc/config.properties``
"""""""""""""""""""""""""""""""""""""""""""""""

> This file enables the coordinator mode, the discovery server, and sets the HTTP port to ``8080``.

Reviewer: ```suggestion
Create the file `coordinator/etc/config.properties` with the following contents. This file enables the coordinator mode, the discovery server, and sets the HTTP port to ``8080``.
```

.. code-block:: properties

   # coordinator/etc/config.properties
   coordinator=true
   node-scheduler.include-coordinator=true
   http-server.http.port=8080
   discovery-server.enabled=true
   discovery.uri=http://localhost:8080

* ``coordinator=true``: Enables the coordinator mode.
* ``discovery-server.enabled=true``: Designates the coordinator as the host for the worker discovery service.

> C. Create ``coordinator/etc/jvm.config``

Reviewer: ```suggestion
3. Create ``coordinator/etc/jvm.config``
```

""""""""""""""""""""""""""""""""""""""""

> These are standard **Java 17** flags for Presto, optimizing the JVM.

Reviewer: ```suggestion
Create the file ``coordinator/etc/jvm.config`` with the following content. These are standard **Java 17** flags for Presto, optimizing the JVM.
```
"optimizing the JVM" ...for what?

Reviewer: > Thanks @steveburnett , I have added this. 

I don't see the text you're highlighting in the file in the branch in this PR. Have you pushed your change to GitHub?

   -XX:G1HeapRegionSize=32M
   -XX:+UseGCOverheadLimit
   -XX:+ExplicitGCInvokesConcurrent
   -XX:+HeapDumpOnOutOfMemoryError
   -XX:+ExitOnOutOfMemoryError
   -Djdk.attach.allowAttachSelf=true
   --add-opens=java.base/java.io=ALL-UNNAMED
   --add-opens=java.base/java.lang=ALL-UNNAMED
   --add-opens=java.base/java.lang.ref=ALL-UNNAMED
   --add-opens=java.base/java.lang.reflect=ALL-UNNAMED
   --add-opens=java.base/java.net=ALL-UNNAMED
   --add-opens=java.base/java.nio=ALL-UNNAMED
   --add-opens=java.base/java.security=ALL-UNNAMED
   --add-opens=java.base/javax.security.auth=ALL-UNNAMED
   --add-opens=java.base/javax.security.auth.login=ALL-UNNAMED
   --add-opens=java.base/java.text=ALL-UNNAMED
   --add-opens=java.base/java.util=ALL-UNNAMED
   --add-opens=java.base/java.util.concurrent=ALL-UNNAMED
   --add-opens=java.base/java.util.concurrent.atomic=ALL-UNNAMED
   --add-opens=java.base/java.util.regex=ALL-UNNAMED
   --add-opens=java.base/jdk.internal.loader=ALL-UNNAMED
   --add-opens=java.base/sun.security.action=ALL-UNNAMED
   --add-opens=java.security.jgss/sun.security.krb5=ALL-UNNAMED

D. Create ``coordinator/etc/node.properties``
"""""""""""""""""""""""""""""""""""""""""""""

> This file sets the node environment and the data directory.

Reviewer: ```suggestion
Create the file ``coordinator/etc/node.properties`` with the following content. This file sets the node environment and the data directory.
```

.. code-block:: properties

   # coordinator/etc/node.properties
   node.id=${ENV:HOSTNAME}
   node.environment=test
   node.data-dir=/var/lib/presto/data

E. Add TPCH Catalog Configuration
"""""""""""""""""""""""""""""""""

> The TPCH catalog enables running test queries against an in-memory dataset.

Reviewer: ```suggestion
Create the file ``coordinator/etc/catalog/tpch.properties`` with the following content. The TPCH catalog enables running test queries against an in-memory dataset.
```

.. code-block:: properties

   # coordinator/etc/catalog/tpch.properties
   connector.name=tpch

Configure the Prestissimo (C++) Worker
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The Worker must be configured to locate the Coordinator or Discovery service and identify itself within the network.

Repeat this step to add more workers, such as ``worker-2``.

A. Create Worker Configuration Directory
""""""""""""""""""""""""""""""""""""""""

.. code-block:: bash

   mkdir -p worker-1/etc/catalog

> B. Create ``worker-1/etc/config.properties``

Reviewer: Please look at what you did in line 50 (2. Create the Coordinator Configuration File) and line 53 (Create the file ``coordinator/etc/config.properties`` with the following contents.). 

Revise these headings the same way: describe the action in the heading, and put the imperative instruction as the first sentence of the text.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

This file orchestrates both the Java Coordinator and the C++ Worker containers. Create the file ``docker-compose.yml`` in your ``~/presto-lab`` directory.

.. code-block:: yaml

   # docker-compose.yml
   services:
     coordinator:
       image: public.ecr.aws/oss-presto/presto:latest
       platform: linux/amd64
       container_name: presto-coordinator
       hostname: coordinator
       ports:
         - "8080:8080"
       volumes:
         - ./coordinator/etc:/opt/presto-server/etc:ro
       restart: unless-stopped

     worker-1:
       image: public.ecr.aws/oss-presto/presto-native:latest
       platform: linux/amd64
       container_name: prestissimo-worker-1
       hostname: worker-1
       depends_on:
         - coordinator
       volumes:
         - ./worker-1/etc:/opt/presto-server/etc:ro
       restart: unless-stopped

     worker-2:
       image: public.ecr.aws/oss-presto/presto-native:latest
       platform: linux/amd64
       container_name: prestissimo-worker-2
       hostname: worker-2
       depends_on:
         - coordinator
       volumes:
         - ./worker-2/etc:/opt/presto-server/etc:ro
       restart: unless-stopped

> * The **coordinator** service uses the standard **Java Presto image** (presto:latest).

Reviewer: remove bold

* The **worker-1** and **worker-2** services use the **Prestissimo (C++ Native) image** (presto-native:latest).
* The setting ``platform: linux/amd64`` is essential for users running on Apple Silicon Macs.
* The ``volumes`` section mounts your local configuration directories (``./coordinator/etc``, ``./worker-1/etc``) into the container's expected path (``/opt/presto-server/etc``).

Step 5: Start the Cluster and Verify
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

A. Start the Cluster
""""""""""""""""""""

Use Docker Compose to start the cluster in detached mode (``-d``).

.. code-block:: bash

   docker compose up -d

B. Verify
"""""""""

1.  **Check the Web UI:** Open the Presto Web UI at http://localhost:8080.

    * *Verification Result:* You should see the UI displaying 3 Active Workers (1 Coordinator and 2 Workers).

> 2.  **Check Detailed Node Status (SQL Query):** Run the following query to check the detailed status and metadata about every node (Coordinator and Workers).

Reviewer: ```suggestion
2.  **Check Detailed Node Status**: Run the following SQL query to check the detailed status and metadata about every node (Coordinator and Workers).
```

The adoption of a C++ execution engine is a significant performance innovation for Presto, offering key advantages for data lake analytics:

* **Massive Performance Boost:** Prestissimo achieves dramatic increases in **CPU efficiency** and reduces query latency by leveraging native C++ execution, **vectorization**, and **SIMD** (Single Instruction, Multiple Data) instructions. Production results have shown fleet sizes shrinking to nearly a third.
* **Eliminates Java GC Issues:** By moving the execution engine outside the JVM, this architecture removes the unpredictable performance spikes and pauses often associated with **Java Garbage Collection (GC)**, resulting in more consistent and stable query times.
* **Explicit Memory Control:** The Velox memory management framework offers **explicit memory accounting** and **arbitration**, providing finer control over resource consumption than the JVM.

II. Prerequisites
-----------------

To follow this tutorial, you need:

* **Docker** installed.
* Basic familiarity with the Terminal and shell commands.

III. Setup Guide
----------------

The recommended directory structure uses ``presto-lab`` as the root directory.

> Step 1: Create a Working Directory

Reviewer: Remove "Step " from all of these numbered headings.

.. code-block:: bash

   mkdir -p ~/presto-lab
   cd ~/presto-lab

Step 2: Configure the Presto Java Coordinator
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The Coordinator requires configuration to define its role, enable the discovery service, and set up a catalog for querying.

A. Create Configuration Directory
"""""""""""""""""""""""""""""""""

.. code-block:: bash

   mkdir -p coordinator/etc/catalog

This command creates the necessary directories for the coordinator and its catalogs.

B. Configure ``coordinator/etc/config.properties``
""""""""""""""""""""""""""""""""""""""""""""""""""

This file enables coordinator mode, the discovery server, and sets the HTTP port to ``8080``.

.. code-block:: properties

   # coordinator/etc/config.properties
   coordinator=true
   node-scheduler.include-coordinator=true
   http-server.http.port=8080
   discovery-server.enabled=true
   discovery.uri=http://localhost:8080

* ``coordinator=true``: Enables the coordinator mode.
* ``discovery-server.enabled=true``: Designates the coordinator as the host for the worker discovery service.

> C. Configure ``coordinator/etc/jvm.config``

Reviewer: Create ``coordinator/etc/jvm.config``? 

As an alternative, you could change the heading to "Configure the JVM" and add a first sentence like
"In this step, configure the JVM by creating the file ``coordinator/etc/jvm.config`` with the following contents:" 

Whatever you choose, be consistent here and with line 62 "B. Configure ``coordinator/etc/config.properties``".

   --add-opens=java.base/javax.security.auth.login=ALL-UNNAMED
   --add-opens=java.base/java.text=ALL-UNNAMED
   --add-opens=java.base/java.util=ALL-UNNAMED
   --add-opens=java.base/java.util.concurrent=ALL-UNNAMED
   --add-opens=java.base/java.util.concurrent.atomic=ALL-UNNAMED
   --add-opens=java.base/java.util.regex=ALL-UNNAMED
   --add-opens=java.base/jdk.internal.loader=ALL-UNNAMED
   --add-opens=java.base/sun.security.action=ALL-UNNAMED
   --add-opens=java.security.jgss/sun.security.krb5=ALL-UNNAMED

D. Configure ``coordinator/etc/node.properties``
""""""""""""""""""""""""""""""""""""""""""""""""

This file sets the node environment and the data directory.

.. code-block:: properties

   # coordinator/etc/node.properties
   node.id=${ENV:HOSTNAME}
   node.environment=test
   node.data-dir=/var/lib/presto/data

E. Add TPCH Catalog Configuration
"""""""""""""""""""""""""""""""""

The **TPCH catalog** enables running test queries against an in-memory dataset.

.. code-block:: properties

   # coordinator/etc/catalog/tpch.properties
   connector.name=tpch

Step 3: Configure the Prestissimo (C++) Worker
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The Worker must be configured to locate the Coordinator/Discovery service and identify itself within the network.

Reviewer: ```suggestion
The Worker must be configured to locate the Coordinator/Discovery service and identify itself within the network. 

Repeat this step to add more workers, such as ``worker-2``.
```

A. Create Worker Configuration Directory
""""""""""""""""""""""""""""""""""""""""

.. code-block:: bash

   mkdir -p worker-1/etc/catalog

B. Configure ``worker-1/etc/config.properties``
"""""""""""""""""""""""""""""""""""""""""""""""

This configuration points the worker to the discovery service running on the coordinator.

.. code-block:: properties

   # worker-1/etc/config.properties
   discovery.uri=http://coordinator:8080
   presto.version=0.288-15f14bb
   http-server.http.port=7777
   shutdown-onset-sec=1
   runtime-metrics-collection-enabled=true

> * ``discovery.uri=http://coordinator:8080``: This uses the **coordinator** service name (as defined in the ``docker-compose.yml`` file) for network communication within Docker.

Reviewer: ```suggestion
* ``discovery.uri=http://coordinator:8080``: This uses the coordinator service name as defined in the ``docker-compose.yml`` file for network communication within Docker.
```
